fix: sync enum trait rendering with enum shape as possible

i10416 opened the pull request:
Previously, rendering result of [enum trait](https://smithy.io/2.0/spec/constraint-traits.html#enum-trait) was much different from that of [enum shape](https://smithy.io/2.0/spec/simple-types.html#enum) in that
1. smithy4s dropped enum variant hints
2. sealed abstract classes generated from enum trait held `smithy.api#enum` hints
3. sealed abstract classes from enum trait didn't have `@param` properties generated.


For example, the following smithy definition generated the Scala code below.

```smithy
@enum([
{
    value: "HEAD",
    name: "HEAD",
    documentation: "head"
},
{
    value: "TAIL",
    name: "TAIL",
    documentation: "tail",
    deprecated: true
}
])

string Coin
```

```scala
// When Coin is generated from enum shape and enum variant has
// documentation, here are `@param` properties for each variant with doc.
sealed abstract class Coin ...

...

object Coin extends Enumeration[Suit] with ...
  ...
  val hints: Hints = Hints(
    smithy.api.Enum(...) // <- When Coin is generated from enum shape, smithy.api.Enum does not exist here.
  )

  /** head */
  case object HEAD extends Coin("HEAD", "HEAD", 0, Hints()) // <- Hints is not empty when Coin is enum shape.
  /** tail  */
  @deprecated
  case object TAIL extends Coin("TAIL", "TAIL", 1, Hints()) // too

```

After this commits, enum variants generated from enum trait also hold hints such as documentation or deprecation. In addition, sealed abstract classes generated from enum trait no longer have redundant `smithy.api#enum hints`.

See https://github.com/disneystreaming/smithy4s/pull/980/files#diff-1374f8e9776053385d67f4bc78d81b77bd7730a72ebd38801479005fba548fa0 for instance.

Note that enum trait and enum shape are not 100% compatible.
For example, enum trait support neither `deprecated$since` nor `deprecated$message`.

i10416 committed:
fix: keep enum rendering as similar as possible

Previously, rendering result of [enum trait](https://smithy.io/2.0/spec/constraint-traits.html#enum-trait)
was much different from that of [enum shape](https://smithy.io/2.0/spec/simple-types.html#enum)
in that smithy4s dropped enum variant hints and sealed abstract classes generated from enum trait held `smithy.api#enum`
hints and sealed abstract classes from enum trait didn't have `@param` properties generated.

For example, the following smithy definition generated the Scala code below.

```smithy
@enum([
{
    value: "HEAD",
    name: "HEAD",
    documentation: "head"
},
{
    value: "TAIL",
    name: "TAIL",
    documentation: "tail",
    deprecated: true
}
])

string Coin
```

```scala
// When Coin is generated from enum shape and enum variant has
// documentation, here are `@param` properties for each variant with doc.
sealed abstract class Coin ...

...

object Coin extends Enumeration[Suit] with ...
  ...
  val hints: Hints = Hints(
    smithy.api.Enum(...) // <- When Coin is generated from enum shape, smithy.api.Enum does not exist here.
  )

  /** head */
  case object HEAD extends Coin("HEAD", "HEAD", 0, Hints()) // <- Hints is not empty when Coin is enum shape.
  /** tail  */
  @deprecated
  case object TAIL extends Coin("TAIL", "TAIL", 1, Hints()) // too

```

After this commits, enum variants generated from enum trait also hold
hints such as documentation or deprecation. In addition, sealed
abstract classes generated from enum trait no longer have
redundant `smithy.api#enum hints`.

Note that enum trait and enum shape are not 100% compatible.

For example, enum trait support neither `deprecated$since` nor
`deprecated$message`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/disneystreaming/smithy4s?pullRequest=980) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/disneystreaming/smithy4s?pullRequest=980) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/disneystreaming/smithy4s?pullRequest=980) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/disneystreaming/smithy4s?pullRequest=980) it.</sub>

i10416 commented on modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:
```diff
@@ -311,31 +311,19 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
-                    value.getValue,
-                    index
-                  ),
-                  hints = Nil
```
This dropped all variants' hints even if there were documentation or deprecation although https://github.com/disneystreaming/smithy4s/pull/980/files#diff-3d87f30f2b03db7c8e012d2c726b1e1a0f1227fe70d2e5c24166a9046c4428a5L337 preserved documentation and deprecation.

i10416 commented on modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:
```diff
@@ -85,6 +85,144 @@ final class RendererSpec extends munit.FunSuite {
+        case Some(code) => code
+      }
+    val classDoc = """/** this is an enum Suit
+                     |  * @param DIAMOND
```
Random thought: would it be better to generate `@param <variant>` even when there's no doc for variant?

i10416 replied:
Hmm, as mentioned https://github.com/disneystreaming/smithy4s/issues/773, `@param` sounds strange for enum variants.

Baccata replied:
Good point, maybe bullet points might be better in this case.

i10416 commented on modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:
```diff
@@ -85,6 +85,144 @@ final class RendererSpec extends munit.FunSuite {
+        case Some(code) => code
+      }
+    val classDoc = """/** this is an enum Suit
+                     |  * @param DIAMOND
```
Random thought: would it be better to generate `@param <variant>` even when there's no doc for variant?

i10416 commented on modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:
```diff
@@ -85,6 +85,144 @@ final class RendererSpec extends munit.FunSuite {
+                     |  * @param HEART
+                     |  *   this is a HEART
+                     |  */""".stripMargin
+    val diamondWithDocRendered =
```
confirm variants holds hints from `@enum` trait

i10416 commented:
Hmm, I missed some corner cases.

i10416 committed:
fix: support unnamed enum trait variants

i10416 committed:
improve: prefer builtin function to build EnumShape

i10416 committed:
fix: reduce diffs

i10416 committed:
docs: update example to demonstrate enum trait hints

kubukoz commented on modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:
```diff
@@ -311,31 +311,55 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
+                        )
+                        .asJavaCollection
+                    )
+                    .asInstanceOf[EnumShape.Builder]
```
**suggestion:** can you use an upcast instead? (`EnumShape.builder().(...): EnumShape.Builder`)

i10416 replied:
I think it gets compile error as `source` and `addTraits` come from `StringShape.Builder`, which is a super class of `EnumShape.Builder`.

kubukoz commented on modules/example/src/smithy4s/example/LowHigh.scala:
```diff
@@ -20,12 +25,12 @@ sealed abstract class LowHigh(_value: String, _name: String, _intValue: Int, _hi
-  val hints: Hints = Hints(
-    smithy.api.Enum(List(smithy.api.EnumDefinition(value = smithy.api.NonEmptyString("Low"), name = Some(smithy.api.EnumConstantBodyName("LOW")), documentation = None, tags = None, deprecated = None), smithy.api.EnumDefinition(value = smithy.api.NonEmptyString("High"), name = Some(smithy.api.EnumConstantBodyName("HIGH")), documentation = None, tags = None, deprecated = None))),
-  )
+  val hints: Hints = Hints.empty
```
**issue:** this doesn't look right, IMO we should keep the traits originally present in the model, regardless if they're duplicated in the Enumeration schema

i10416 replied:
I found `fromStringShape` in software.amazon.smithy.model.shapes.EnumShape.java drops enum trait from `StringShape`. Do we still need to keep it? 


```java
    public static Optional<EnumShape> fromStringShape(StringShape shape, boolean synthesizeNames) {
        if (shape.isEnumShape()) {
            return Optional.of((EnumShape) shape);
        }

        if (!shape.hasTrait(EnumTrait.ID)) {
            return Optional.empty();
        }
        StringShape stringWithoutEnumTrait = shape.toBuilder().removeTrait(EnumTrait.ID).build();
        Builder enumBuilder = EnumShape.builder();
        stringWithoutEnumTrait.updateBuilder(enumBuilder);
        ...
```

i10416 replied:
https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/shapes/EnumShape.java#L131

kubukoz replied:
I guess there's no "right thing to do", it'll be a judgement call - do we try to do what Smithy does in `upgradeEnums` (which is an optional transformation that uses the code path you mentioned), or do we try to make our representation as close to the original as possible?

I think for support of Smithy 1.x, we usually did the latter.

i10416 replied:
Okey, I will keep enum trait information in generated hints.

Baccata replied:
A little late to the party but I think it's fine to omit this trait. All semantic information is recoverable from the Schema, and this trait just bloats the generated code (and therefore bytecode) for no reason. 

If there was a reasonably good reason for advocating for its presence, I'd be inclined to keep it, but in the absence of such an argument, let's remove the generated `enum` hint. 

PS : sorry for the churn.

i10416 committed:
improve: prefer builtin name synthesize to custom one

kubukoz commented:
Request: please avoid force pushes on PRs that have already been seen :) it's fine to keep "bad" commits on the branches, as we will squash them anyway.

i10416 commented:
So sorry for that 🙇

i10416 committed:
fix: preserve enum trait information in generated hints

i10416 commented on modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:
```diff
@@ -311,31 +311,66 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
+                case Some(shape) =>
+                  shape
+                    .toBuilder()
+                    .addTrait(enumTrait)
```
restore `EnumTrait` dropped by `fromStringShape`

Baccata replied:
See my comment here : https://github.com/disneystreaming/smithy4s/pull/980#discussion_r1214054302

i10416 commented on modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:
```diff
@@ -311,31 +311,66 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
+                  }
+                  EnumShape
+                    .builder()
+                    .addTrait(enumTrait)
```
restore `EnumTrait` dropped by `fromStringShape`

Baccata replied:
See my comment here : https://github.com/disneystreaming/smithy4s/pull/980#discussion_r1214054302

i10416 committed:
update: drop redundant enum trait

drop redundant enum trait from generated code as information can be restored from schema as
mentioned in https://github.com/disneystreaming/smithy4s/pull/980#discussion_r1214054302

i10416 committed:
chore: restore traits except enum trait

i10416 committed:
tryfix: avoid multiline value in assertion

I guess possible reason for assertion failure on CI is a difference of line separator between systems though
I'm not sure why assertion for `haertWithDeprecationAndDocRendered` fails on CI while it passes on my local machine.

i10416 commented on modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:
```diff
@@ -85,6 +85,192 @@ final class RendererSpec extends munit.FunSuite {
+      "variants generated from `@enum` trait must hold hints, but deprecation hint cannot support message and since properties."
+    )
+    assert(
+      !definition.contains("smithy.api.Enum"),
```
updated: generated shape doesn't have redundant metadata.

i10416 commented on modules/codegen/test/src/smithy4s/codegen/internals/RendererSpec.scala:
```diff
@@ -85,6 +85,192 @@ final class RendererSpec extends munit.FunSuite {
+    val diamondWithDocRendered =
+      """case object DIAMOND extends Suit("DIAMOND", "DIAMOND", 0, Hints(smithy.api.Documentation("this is a DIAMOND")))"""
+    // NOTE: enum trait and enum shape are not 100% compatible. For example, enum trait doesn't support deprecated$since and deprecated$message.
+    val haertWithDeprecationAndDocRendered = List(
```
Here, I use List instead of single String because multiline assertion somehow failed on CI but worked on my local machine.

Baccata commented:
Thank you !
